cs/commandstation/UpdateProcessor.hxx

```diff
@@ -85,22 +109,40 @@ class UpdateProcessor : public StateFlow<Buffer<dcc::Packet>, QList<1> >,
   Action entry() OVERRIDE;
 
  private:
-  // Place where we forward the packets filled in.
+  struct SourceState {
+    /// Stores the last time we sent a packet to a given loco. Suppresses
+    /// refresh packet if this time is too recent to avoid confusing DCC
+    /// decoders.
+    long long lastPacketTime_;
+    /// Priority of packet source. This also encodes whether this is an
+    /// exclusive packet source.
+    unsigned priority_;
+  };
+
+  /// @return if we have an exclusive source.
+  bool has_exclusive() { return exclusiveIndex_ != NO_EXCLUSIVE; }
+
+  /// Place where we forward the packets filled in.
   dcc::PacketFlowInterface* trackSend_;
 
-  // Holds the list of train nodes that have reported a change. We will always
-  // take a train node from this list first before starting background refresh.
+  /// Holds the list of train nodes that have reported a change. We will always
+  /// take a train node from this list first before starting background refresh.
   QList<1> priorityUpdates_;
 
-  // Packet sources to ask about refreshing data periodically.
+  /// Packet sources to ask about refreshing data periodically.
   vector<dcc::PacketSource*> refreshSources_;
 
-  // Stores the last time we sent a packet to a given loco. Suppresses refresh
-  // packet if this time is too recent to avoid confusing DCC decoders.
-  map<dcc::PacketSource*, long long> lastPacketTime_;
+  /// Stores additional data about each packet source.
+  map<dcc::PacketSource*, SourceState> packetSourceStates_;
 
-  // Which is the next guy on the refresh source list to add.
-  unsigned nextRefreshIndex_ : 16;
+  /// Which is the next guy on the refresh source list to add.
+  unsigned nextRefreshIndex_ : 15;
+  /// The highest priority refresh index, if we have a exclusive index.
+  unsigned exclusiveIndex_ : 15;
+  /// This is stored in the exclusiveIndex_ when we have no exclusive refresh
+  /// source.
+  static constexpr unsigned NO_EXCLUSIVE = 0x7FF;
+  /// Non-zero if we have any refresh sources.
   unsigned hasRefreshSource_ : 1;
```

Comment:
I don't think it matters for packing, but move the this single bit above the constexpr so that it is together with the other bit definitions.

Reply:
done